### PR TITLE
#21: Added JsonRequest and JsonResponseAssertions #fromModel() utility methods.

### PR DESCRIPTION
**Pull request by FrelliBB**
Related issue: https://github.com/bastion-dev/Bastion/issues/21

### CODE REVIEW

File: src/main/java/rocks/bastion/core/json/JsonRequest.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package rocks.bastion.core.json;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
```
Comment (KPull) [inline src/main/java/rocks/bastion/core/json/JsonRequest.java]: This `import` statement is incorrect. We don't want to see any reference to Jackson at the level of Requests.

File: src/main/java/rocks/bastion/core/json/JsonRequest.java [hunk 2]
```diff
@@ -429,6 +430,78 @@ public static JsonRequest deleteFromTemplate(String url, String jsonTemplateSour
...
+     * Construct an HTTP request, using the PATCH method, containing a JSON body that is serialized from the given object. Initially, the request
+     * will have the "application/json" HTTP header and no other additional headers and no query parameters. It will
+     * also have a descriptive name which is generated by combining the HTTP method with the URL.
+     *
+     * @param url    The URL to send this request on
+     * @param model  The object to serialize into JSON text as the body content for this request
+     * @return An HTTP request containing the serialized model as the JSON body text
+     * @throws JsonSerializationException Thrown if an unexpected exeption occurs during serialization
+     */
+    public static JsonRequest patchFromModel(String url, Object model) throws JsonProcessingException {
```
Comment (KPull) [inline src/main/java/rocks/bastion/core/json/JsonRequest.java]: The exception in these `throws` statement is incorrect. You want to throw `JsonSerializationException` which you list in the JavaDocs.

File: src/main/java/rocks/bastion/core/json/JsonSerializationException.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package rocks.bastion.core.json;
+
+import static java.lang.String.format;
+
+/**
+ * Indicates that an object could not be serialized due to an unforeseen error.
+ */
+public class JsonSerializationException extends RuntimeException {
```
Comment (KPull) [inline src/main/java/rocks/bastion/core/json/JsonSerializationException.java]: I'm not sure about this, but do you want to try and start marking these classes `final` as these classes are not meant to be extended?

File: src/main/java/rocks/bastion/core/json/JsonSerializer.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package rocks.bastion.core.json;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * Serializes an {@link Object} into its String representation.
+ */
+class JsonSerializer {
```
Comment (KPull) [inline src/main/java/rocks/bastion/core/json/JsonSerializer.java]: This is another class which is eligible for being `final`. Feel free to discuss this further as I'm not sure about this.

File: src/main/java/rocks/bastion/core/json/JsonSerializer.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    private final Object source;
+    private final ObjectMapper objectMapper;
+
+    JsonSerializer(Object source) {
```
Comment (KPull) [inline src/main/java/rocks/bastion/core/json/JsonSerializer.java]: This class is package-private, which I agree with and we can leave it as it is. But I believe that the methods within it should be `public`: the reasoning behind this is that, once you have a reference to `JsonSerializer`, then you want to execute methods on its `public` API. Methods which are package-private usually have a different design objective behind them (for example, they are meant to be overridden by classes within the package).

File: src/test/java/rocks/bastion/core/assertions/JsonResponseAssertionsTest.java [hunk 6]
```diff
@@ -14,6 +16,28 @@
...
 public class JsonResponseAssertionsTest {
 
+    @Test(expected = AssertionError.class)
+    public void fromModel_differentValue_assertShouldFail() {
+        try {
+            final Sushi expectedModel = Sushi.newSushi().id(999).name("Salmon Nigiri").type(Sushi.Type.NIGIRI).price(50L).build();
+            final ModelResponse<String> response = TestModelResponse.prepare("{\"id\":1,\"name\":\"Salmon Nigiri\",\"type\":\"NIGIRI\",\"price\":50}");
+            final JsonResponseAssertions assertions = JsonResponseAssertions.fromModel(200, expectedModel);
+            assertions.execute(200, response, response.getModel());
+        } catch(AssertionError assertionError) {
```
Comment (KPull) [inline src/test/java/rocks/bastion/core/assertions/JsonResponseAssertionsTest.java]: We can make this test cleaner by using AssertJ's `assertThatThrownBy()`. You can then even remove the `expected` parameter from within the `@Test` annotation as AssertJ will take care of things by itself.

File: src/main/java/rocks/bastion/core/json/JsonResponseAssertions.java [hunk 7]
```diff
@@ -109,6 +109,22 @@ public static JsonResponseAssertions fromTemplate(int expectedStatusCode, String
...
+     *
+     * The comparison is made by serializing the provided expected model and comparing the resultant JSON String to the response
+     * body's content as with {@link #fromString(int, String)}.
+     *
+     * @param expectedStatusCode The expected HTTP status code
+     * @param expectedModel      The model which will be deserialized into the expected JSON object.
+     * @return An assertions object for use with the {@link rocks.bastion.core.builder.AssertionsBuilder#withAssertions(Assertions)} method
+     */
+    public static JsonResponseAssertions fromModel(int expectedStatusCode, Object expectedModel) {
+        final String expectedJson = new JsonSerializer(expectedModel).serialize();
```
Comment (KPull) [inline src/main/java/rocks/bastion/core/json/JsonResponseAssertions.java]: These `final` qualifiers are unnecessary. You need to also remove them from elsewhere in this pull request.

**Comment by KPull** [comment]
@FrelliBB, with regards to a configurable `ObjectMapper`: I fully agree with this. Having said that, this issue should not stop us from merging this pull request since exposing functionality for configuring the `ObjectMapper` can be tackled separately (and won't really conflict with the changes here).

**Comment by FrelliBB** [comment]
Changes applied @KPull

**Pull request merged**